lib/charms/opensearch/v0/helpers/conf_setter.py

```diff
@@ -0,0 +1,297 @@
+from ruamel.yaml.comments import CommentedSet
+
+# The unique Charmhub library identifier, never change it
+LIBID = "f4bd9c1dad554f9ea52954b8181cdc19"
+LIBAPI = 0
+LIBPATCH = 0
+
+
+class OutputType(Enum):
+    """Enum representing the output type of a write operation."""
+
+    file = "file"
+    obj = "obj"
+    console = "console"
+    all = "all"
+
+    def __str__(self):
+        """String representation of enum value."""
+        return self.value
+
+
+class YamlConfigSetter:
+    """Utility class for updating YAML config, supporting diverse object types and nestedness.
+
+    conf_setter = ConfigSetter()
+    put("file.yml", "a.b", "new_name")
+    put("file.yml", "a.b/c.obj/key3/key1.a/obj", {"a": "new_name_1", "b": ["hello", "world"]})
+    put("file.yml", "a.b/c.arr.simple/[0]", "hello")
+    put("file.yml", "a.b/c.arr.complex/[name:name1]", {"a": "new_name_1", "b": ["hello", "world"]})
+    put("file.yml", "a.b/c.arr.complex/[name:name5]/val/key", "new_val25")
+    put("file.yml", "a.b/c.arr.complex/[0]/val", "complex2_updated")
+    put("file.yml", "a.b/c.arr.complex/[0]/new_val/new_sub_key", "updated")
+    """
+
+    def __init__(self, base_path: str = None):
+        """base_path: if set, where to look for files relatively on "load/put/delete" methods."""
+        self.yaml = YAML()
+        self.base_path = self.__clean_base_path(base_path)
+
+    def load(self, config_file: str) -> Dict[str, any]:
+        """Load the content of a YAML file."""
+        path = f"{self.base_path}{config_file}"
+
+        if not exists(path):
+            raise FileNotFoundError(f"{path} not found.")
+
+        random_id = uuid.uuid4().hex
+        with open(path, "r") as f:
+            lines = f.readlines()
+
+            doc_empty_or_full_comments = True
+            for line in lines:
+                if not line.startswith("#") and line.strip() != "":
+                    doc_empty_or_full_comments = False
+                    break
+
+        if doc_empty_or_full_comments:
+            with open(path, "a") as f:
+                f.write(f'\n{random_id}: "{random_id}"')
+
+        with open(path, mode="r") as f:
+            data = self.yaml.load(f)
+            if doc_empty_or_full_comments:
+                del data[random_id]
+
+        if doc_empty_or_full_comments:
+            self.__dump(data, OutputType.file, path)
+
+        return data
+
+    def put(
+        self,
+        config_file: str,
+        key_path: str,
+        val: any,
+        sep="/",
+        output_type: OutputType = OutputType.file,
+        inline_array: bool = False,
+        output_file: str = None,
+    ) -> Dict[str, any]:
+        """Add or update the value of a key (or content of array at index / key) if it exists."""
+        data = self.load(config_file)
+
+        self.__deep_update(data, key_path.split(sep), val)
+
+        if inline_array:
+            data = self.__inline_array_format(data, key_path.split(sep), val)
+
+        self.__dump(
```

Comment:
After talking this over on MM, I see why you took this approach - avoiding rewrites and restarts is important on PGB because you have to restart the entire application, so my proposed approach suits better, but for opensearch I understand your approach.